ADAL/tests/unit/ADUserInformationTests.m

```diff
@@ -41,14 +41,141 @@ - (void)tearDown
+    [rhs setValue:@"id 2" forKey:@"uniqueId"];
+    
+    XCTAssertNotEqualObjects(lhs, rhs);
+}
+
+- (void)testIsEqual_whenUniqueIdIsEqual_shouldReturnTrue
+{
+    ADUserInformation *lhs = [self createEmptyUserInformation];
+    [lhs setValue:@"id 1" forKey:@"uniqueId"];
+    ADUserInformation *rhs = [self createEmptyUserInformation];
+    [rhs setValue:@"id 1" forKey:@"uniqueId"];
+    
+    XCTAssertEqualObjects(lhs, rhs);
+}
+
+- (void)testIsEqual_whenUserIdDisplayableIsNotEqual_shouldReturnFalse
+{
+    ADUserInformation *lhs = [self createEmptyUserInformation];
+    [lhs setValue:@NO forKey:@"userIdDisplayable"];
+    ADUserInformation *rhs = [self createEmptyUserInformation];
+    [rhs setValue:@YES forKey:@"userIdDisplayable"];
+    
+    XCTAssertNotEqualObjects(lhs, rhs);
+}
+
+- (void)testIsEqual_whenUserIdDisplayableIsEqual_shouldReturnTrue
+{
+    ADUserInformation *lhs = [self createEmptyUserInformation];
+    [lhs setValue:@YES forKey:@"userIdDisplayable"];
+    ADUserInformation *rhs = [self createEmptyUserInformation];
+    [rhs setValue:@YES forKey:@"userIdDisplayable"];
+    
+    XCTAssertEqualObjects(lhs, rhs);
+}
+
+- (void)testIsEqual_whenUserIdIsNotEqual_shouldReturnFalse
+{
+    ADUserInformation *lhs = [self createEmptyUserInformation];
+    [lhs setValue:@"asd" forKey:@"userId"];
+    ADUserInformation *rhs = [self createEmptyUserInformation];
+    [rhs setValue:@"qwe" forKey:@"userId"];
+    
+    XCTAssertNotEqualObjects(lhs, rhs);
+}
+
+- (void)testIsEqual_whenUserIdIsEqual_shouldReturnTrue
+{
+    ADUserInformation *lhs = [self createEmptyUserInformation];
+    [lhs setValue:@"qwe" forKey:@"userId"];
+    ADUserInformation *rhs = [self createEmptyUserInformation];
+    [rhs setValue:@"qwe" forKey:@"userId"];
+    
+    XCTAssertEqualObjects(lhs, rhs);
+}
+
+#pragma mark - userInformationWithIdToken
+
+- (void)testUserInformationWithIdToken_whenUpnContainsSpaces_shouldReturnTrimmedUserId
+{
+    NSDictionary *part1Claims = @{ @"typ" : @"JWT", @"alg" : @"none" };
+    NSString *p1 = [NSString adBase64UrlEncodeData:[NSJSONSerialization dataWithJSONObject:part1Claims options:0 error:nil]];
+    NSDictionary *idtokenClaims = @{@"upn" : @"     eric_cartman@contoso.com           ",};
+    NSString *p2 = [NSString adBase64UrlEncodeData:[NSJSONSerialization dataWithJSONObject:idtokenClaims options:0 error:nil]];
+    NSString *idtoken = [NSString stringWithFormat:@"%@.%@", p1, p2];
+    
+    ADUserInformation *userInfo = [ADUserInformation userInformationWithIdToken:idtoken error:nil];
+    
+    ADAssertStringEquals(userInfo.userId, @"eric_cartman@contoso.com");
+}
+
+#pragma mark -
+
 - (void) testCopy
 {
     ADUserInformation* userInfo = [self adCreateUserInformation:@"eric_cartman@contoso.com"];
     XCTAssertNotNil(userInfo);
     
     ADUserInformation* copy = [userInfo copy];
     XCTAssertNotNil(copy);
-    XCTAssertNotEqualObjects(copy, userInfo);
+    XCTAssertEqualObjects(copy, userInfo);
```

Comment:
We still need to check them in order to be 100% sure that copy method works properly. Otherwise if we just compare 2 instances of userInfo and our isEqual method is broken this test might give you false positive.